### DISCUSSION TITLE
Why only BGR formats for FrameOutput?

### DISCUSSION THREAD

**Discussion by acras01** [kokorin/Jaffree#393] [Q&A]
Hello! I'm new here so sorry for newbie question. Why does lib suggest only BGR24/BGRA image formats? In our application we need ARGB, so we have to perform the byte conversion on-the-fly. Any restrictions by ffmpeg itself? Thank you!

**Comment by kokorin** [comment]
Jaffree uses NUT format to read/write `BufferedImage`. NUT format supports any pixel format which ffmpeg supports, so there is no limitation in ffmpeg.

But the problem is that conversion between pixel formats requires a lot of CPU, so Jaffree opts for no pixel format conversion. Instead it reads raw bytes and constructs `BufferedImage` with no byte manipulations.
I found only 2 formats that can be converted to BufferedImage: `argb` and `bgr24`.

Check [ImageFormats](https://github.com/kokorin/Jaffree/blob/master/src/main/java/com/github/kokorin/jaffree/ffmpeg/ImageFormats.java) for details. To support new pixel format one just needs to corresponding `ImageFormat` implementation.

**Reply by acras01** [reply]
> I found only 2 formats that can be converted to BufferedImage: `argb` and `bgr24`.

`argb`? I saw bgra in source code. Sorry, but I do not understand why only these two as 
`public BufferedImage (ColorModel cm,
                          WritableRaster raster,
                          boolean isRasterPremultiplied,
                          Hashtable<?,?> properties)`
doesn't seem to have limitations in pixel format

**Reply by acras01** [reply]
> To support new pixel format one just needs to corresponding `ImageFormat` implementation.

Am I supposed to make a fork with all needed edits to achieve my goal?

**Reply by kokorin** [reply]
> argb? I saw bgra in source code

Could you point to that piece of code?

> public BufferedImage (ColorModel cm, WritableRaster raster, boolean isRasterPremultiplied, Hashtable<?,?> properties)
doesn't seem to have limitations in pixel format

It's a constructor, you have to dive deeper. As I said: I found only 2 formats that can be converted to `BufferedImage`: argb and bgr24.
If you found another formats - I will merge changes.

> Am I supposed to make a fork with all needed edits to achieve my goal?

That's exactly how open source works. And, actually, you don't even need a fork. `ImageFormat` is an interface which has 2 implementations in `ImageFormats`. So in you project you could implement `ImageFormat` and use it. But to be honest, I would be interested in your findings.

**Reply by kokorin** [reply]
Could also clarify what pixel format (or image format) you need?

**Reply by acras01** [reply]
> Could also clarify what pixel format (or image format) you need?

I need argb, but `withConsumerAlpha` gives me bgra

**Reply by acras01** [reply]
> Could you point to that piece of code?

https://github.com/kokorin/Jaffree/blob/b503d1fb3e5ac54f6dc1f315321cb8d696818d9c/src/main/java/com/github/kokorin/jaffree/ffmpeg/FrameOutput.java#L180

**Reply by acras01** [reply]
> So in you project you could implement `ImageFormat` and use it.

I can't do it, cause appropriate method is protected 
`protected static FrameOutput withConsumer(final FrameConsumer consumer,
                                              final ImageFormat imageFormat)`

**Reply by kokorin** [reply]
Sorry, my bad, I missed `argb` with `abgr`. 
Let me clarify this once again: I didn't find any other format (except `abgr` and `bgr24`) which can be converted to `BufferedImage` without byte manipulation. `ImageFormats` works quite fast because it doesn't construct `BufferedImage` and fill it pixel but pixel, but it instead uses byte array (read from ffmpeg output) and creates `BufferedImage` with byte array as source.

> I can't do it, cause appropriate method is protected

I agree, this method, probably, should be public. But subclass has access to this method, right?

> I need argb

Could you provide code snippet, which creates `BufferedImage` from byte array in `argb` without byte manipulations?

**Reply by kokorin** [reply]
Check email
